src/notebooks/controllers/kernelSource/notebookKernelSourceSelector.ts

```diff
@@ -182,3 +451,33 @@ function groupBy<T>(data: ReadonlyArray<T>, compare: (a: T, b: T) => number): T[
 function compareIgnoreCase(a: string, b: string) {
     return a.localeCompare(b, undefined, { sensitivity: 'accent' });
 }
+
+function waitForNotebookControllersCreationForServer(
+    serverId: string,
+    controllerRegistration: IControllerRegistration
+) {
+    if (
+        controllerRegistration.all.find(
+            (connection) =>
+                (connection.kind === 'connectToLiveRemoteKernel' || connection.kind === 'startUsingRemoteKernelSpec') &&
+                connection.id === serverId
+        )
+    ) {
+        return;
+    }
+
+    return new Promise<void>((resolve) => {
+        controllerRegistration.onChanged((e) => {
+            for (let controller of e.added) {
+                if (
+                    controller.connection.kind === 'connectToLiveRemoteKernel' ||
+                    controller.connection.kind === 'startUsingRemoteKernelSpec'
+                ) {
+                    if (controller.connection.serverId === serverId) {
+                        resolve();
+                    }
+                }
```

Comment:
please add the event listeners to the disposables, do that they get disposed appropriately

Reply:
Good catch, I changed to dispose the listener once it resolves.

Reply:
Currently we didn't have proper cancelation handling yet, there might be pending promising hanging around when the quick pick is dismissed and restarted. I'll send another PR to have this handled properly.